src/Frontend/Core.hs

```diff
@@ -274,7 +274,14 @@ class Page p => FormPage p where
 
     -- | The form action used in form generation
     formAction :: p -> P.Main
-    -- | Calculates a redirect address from the given page
+    -- | Calculates a redirect address from the given page.
+    --
+    -- Currently, 'FormPage' forces you to redirect after a form has been processed successfully.
+    -- This may not always be what you want.  For instance, a `preview` button of a text field
+    -- containing markdown does not change the state of the server, and multiple re-submissions are
+    -- a feature, not a problem.  For those cases, we may need to extend this part of the API in the
+    -- future.  We could make the result type a @Maybe a@, where 'Nothing' means redirect, and
+    -- 'Just' means not redirect.  Or we could add a more flexible variant of the 'FormPage' class.
```

Comment:
[Context #398 and #83]

We could also say that these extra submissions are not using FormH but PostH and GetH. In your example of the preview button, my view is that it should not be a post, even though I see the point of using the form data.